### DISCUSSION TITLE
How can I directly return an error (for example in an AsyncValue) out of a FutureProvider?

### DISCUSSION THREAD

**Discussion by timleg002** [rrousselGit/riverpod#1550] [Q&A] [answered]
Something like this:
```dart
final tvApiUserInfo = FutureProvider.autoDispose((ref) async {
  final res = await tvApi.getUserInfo();

  if(res.isOk()) {
    return AsyncValue.data(res.unwrap()); // the user info object
  } else {
    return AsyncValue.error(res.unwrapErr()); // a separate error type class containing an enum & a string containing the message
  }
});
```

I want to trigger the error callback of the `.when()` call myself.

But when I `ref.watch` the provider, I only get an `AsyncValue<AsyncValue<dynamic>>`. I'm a beginner with state management in Flutter (years of only using `setState` produced some horrid, unmaintainable code), so I may not know the best practices, but I'm eager to learn! Thanks for the help :)

**Comment by rrousselGit** [comment, marked as answer]
You may want to throw

```dart
final tvApiUserInfo = FutureProvider.autoDispose((ref) async {
  final res = await tvApi.getUserInfo();

  if(res.isOk()) {
    return res.unwrap();
  } else {
    throw res.unwrapErr();
  }
});
```

**Reply by timleg002** [reply]
Thank you, it's working now. Although in future, I'd prefer APIs that would do something akin to what I mentioned (`AsyncValue.data()`, `AsyncValue.error()`, ...)

**Comment by AhmedLSayed9** [comment]
I've to use StateNotifierProvider for this reason.
I'm throwing the exception at the data layer and catch it at the domain layer (at the Repo) then I return Failure or data using `Either`. Failure is a custom class so I can't throw it as @rrousselGit mentioned. I've to use AsyncValue with StateNotifierProvider and set error state like this:

```
final result = await callApi();
result.fold(
      (failure) {
        state = AsyncError(failure, stackTrace: failure.stackTrace);
      },
      (data) {
        state = const AsyncData(data);
      },
    )
```

**Reply by mregnauld** [reply]
And for that reason, `stackTrace` should be an optional parameter of `AsyncError`, in my opinion.

**Reply by rrousselGit** [reply]
@mregnauld No, pass StackTrace.current then

**Reply by AhmedLSayed9** [reply]
I was able to solve this by using `throwWithStackTrace` method as following:
```
Error.throwWithStackTrace(failure, failure.stackTrace);
```
I think StackTrace being required is an advantage.

**Reply by AhmedLSayed9** [reply]
Recently, I've stopped using `Either` pattern because it's a bit redundant with Riverpod as `AsyncValue` is doing the same job.
